### PR TITLE
Crunch

### PR DESCRIPTION
**Pull request by Youjose**

### CODE REVIEW

**Comment by K0lb3** [comment]
So, the relevant notes to do:

1. fixing the compiler warnings (errors in github action build)
`error: variable does not need to be mutable
   --> src/unitycrunch/crn_unpacker.rs:551:25
    |
551 |                     let mut buffer = &mut self.m_block_buffer[x];
    |                         ----^^^^^^
    |                         |
    |                         help: remove this `mut`
    |
    = note: `-D unused-mut` implied by `-D warnings`

error: variable does not need to be mutable
   --> src/unitycrunch/crn_unpacker.rs:614:25
    |
614 |                     let mut buffer = &mut self.m_block_buffer[x];
    |                         ----^^^^^^
    |                         |
    |                         help: remove this `mut`

error: variable does not need to be mutable
   --> src/unitycrunch/crn_unpacker.rs:693:25
    |
693 |                     let mut buffer = &mut self.m_block_buffer[x];
    |                         ----^^^^^^
    |                         |
    |                         help: remove this `mut`

error: variable does not need to be mutable
   --> src/unitycrunch/crn_unpacker.rs:756:25
    |
756 |                     let mut buffer = &mut self.m_block_buffer[x];
    |                         ----^^^^^^
    |                         |
    |                         help: remove this `mut`

error: variable does not need to be mutable
   --> src/unitycrunch/crn_unpacker.rs:831:25
    |
831 |                     let mut buffer = &mut self.m_block_buffer[x << 1];
    |                         ----^^^^^^
    |                         |
    |                         help: remove this `mut`

error: variable does not need to be mutable
   --> src/unitycrunch/crn_unpacker.rs:933:25
    |
933 |                     let mut buffer = &mut self.m_block_buffer[x << 1];
    |                         ----^^^^^^
    |                         |
    |                         help: remove this `mut`
`

2. see following comments

File: src/lib.rs [hunk 1]
```diff
@@ -83,10 +88,15 @@ mod atc;
 mod bcn;
 mod etc;
 mod pvrtc;
+mod crunch;
```
Comment (K0lb3) [inline src/lib.rs:91]: As crunch uses alloc,
it should be locked behind a `#[cfg(feature = "alloc")]` to allow complete no-std deployments.

File: src/crunch.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+fn crunch_unpack_level<'vec>(data: &[u8], data_size: u32, level_index: u32) -> Result<CrunchDecodeHandler, &'static str> {
+    let mut tex_info: crn_decomp::crn_texture_info = crn_decomp::crn_texture_info::default();
+    if tex_info.crnd_get_texture_info(data, data_size) == false {
+        return Err("Invalid crunch texture encoding.");
+    }
+    let mut p_context: crn_unpacker::crn_unpacker<'_> = match crn_decomp::crnd_unpack_begin(&data, data_size){
+        Ok(p_context) => p_context,
+        Err(res) => return Err(res)
+    };
+    let width = core::cmp::max(1, tex_info.m_width >> level_index);
```
Comment (K0lb3) [inline src/crunch.rs:25]: Importing max via `use core::cmp::max` would make this easier to read.

File: src/crunch.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    match handler.format{
+        crnlib::crn_format::cCRNFmtDXT1 => bcn::decode_bc1(&handler.dxt_data, width, height, image),
+
+        crnlib::crn_format::cCRNFmtDXT5 |
+        crnlib::crn_format::cCRNFmtDXT5_CCxY |
+        crnlib::crn_format::cCRNFmtDXT5_xGBR |
+        crnlib::crn_format::cCRNFmtDXT5_AGBR |
+        crnlib::crn_format::cCRNFmtDXT5_xGxR => bcn::decode_bc3(&handler.dxt_data, width, height, image),
+
+        // crnlib::crn_format::cCRNFmtDXT5A => ???,
```
Comment (K0lb3) [inline src/crunch.rs:57]: I guess we could think about adding support for this as well.

According to [this reply to XeNTaX/[DXT5A (Xbox360)](https://forum.xentax.com/viewtopic.php?p=88395#p88395) the format of DXT5A is ATI1 aka BC4.

So ``crnlib::crn_format::cCRNFmtDXT5A => bcn::decode_bc4(&handler.dxt_data, width, height, image),``
should work

File: src/crunch/crn_consts.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
+#![allow(non_upper_case_globals)]
+#![allow(dead_code)]
+
+// Moved Consts from structs.
+pub const cCRNHeaderMinSize: u32 = 62;
```
Comment (K0lb3) [inline src/crunch/crn_consts.rs:5]: It might make sense to convert most of the values here to usize,
considering that most of them are likely used in the context of indexing.

File: src/crunch/crn_decomp.rs [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        val <<= 8 * (4 - N);
+        for i in 0..N{
+            self.m_buf[i] = (val >> 24) as u8;
+            val <<= 8; 
+        }
+        return self;
+    }
+
+    pub fn cast_to_uint(&mut self) -> u32{
+        return match N {
```
Comment (K0lb3) [inline src/crunch/crn_decomp.rs:166]: This could be replaced with

```rust
 `        return match N {
            1 => self.m_buf[0] as u32,
            2 => u16::from_be_bytes(self.m_buf),
            3 => self.m_buf[0] << 16 | u16::from_be_bytes(self.m_buf[1..=2]),
            4 => u32::from_be_bytes(self.m_buf)
        }
        ````
        to make it easier to read

File: src/crunch/crn_decomp.rs [hunk 6]
```diff
@@ -0,0 +1,347 @@
+#![allow(non_snake_case)]
+#![allow(non_camel_case_types)]
+#![allow(dead_code)]
```
Comment (K0lb3) [inline src/crunch/crn_decomp.rs:3]: Instead of allowing dead code,
it might be better to simply comment it out.

File: src/crunch/crn_decomp.rs [hunk 7]
```diff
@@ -0,0 +1,347 @@
+#![allow(non_snake_case)]
```
Comment (K0lb3) [inline src/crunch/crn_decomp.rs:1]: It might be better to simply convert all non xyz case types to the normal Rust convention.
-> simple refractoring via any IDE

File: src/crunch/crn_static_huffman_data_model.rs [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+            self.m_table_max_code = 0;
+        }
+        // sentinels
+        self.m_max_codes[cMaxExpectedCodeSize as usize] = u32::MAX;
+        self.m_val_ptrs[cMaxExpectedCodeSize as usize] = 0xFFFFF;
+
+        self.m_table_shift = 32 - self.m_table_bits;
+        return true;
+    }
+    #[inline]
```
Comment (K0lb3) [inline src/crunch/crn_static_huffman_data_model.rs:159]: something something, formatting / new line missing :P

File: src/crunch/crn_static_huffman_data_model.rs [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    pub m_total_syms: u32,
+    pub m_code_sizes: alloc::vec::Vec<u8>,
+    pub m_pDecode_tables: decoder_tables
+}
+
+impl static_huffman_data_model{
+    pub fn clear(&mut self){
+        *self = static_huffman_data_model::default();
+    }
+    #[allow(dead_code)]
```
Comment (K0lb3) [inline src/crunch/crn_static_huffman_data_model.rs:183]: ...

File: src/crunch/crn_unpacker.rs [hunk 10]
```diff
@@ -0,0 +1,742 @@
...
+
+            _ => false
+        };
+        if res == false {
+            return Err("Invalid or unsupported Crunch encoding format.");
+        }
+        self.m_codec.stop_decoding();
+        return Ok(ret);
+    }
+    pub fn unpack_dxt1(&mut self, pDst: &mut [u8], row_pitch_in_bytes: u32, blocks_x: u32, blocks_y: u32, chunks_x: u32, chunks_y: u32) -> bool{
```
Comment (K0lb3) [inline src/crunch/crn_unpacker.rs:383]: bool -> Result

File: src/crunch/crn_unpacker.rs [hunk 11]
```diff
@@ -0,0 +1,742 @@
...
+                        }
+                    }
+                    block_dst = (block_dst as i32 + block_delta) as usize;
+                }
+                row_dst += (row_pitch_in_bytes * 2) as usize;
+            }
+        }
+        return true;
+    }
+    pub fn unpack_dxt5(&mut self, pDst: &mut [u8], row_pitch_in_bytes: u32, blocks_x: u32, blocks_y: u32, chunks_x: u32, chunks_y: u32) -> bool{
```
Comment (K0lb3) [inline src/crunch/crn_unpacker.rs:479]: bool -> Result

File: src/crunch/crn_unpacker.rs [hunk 12]
```diff
@@ -0,0 +1,742 @@
...
+                        pd_dst >>= 2;
+                    }
+                    block_dst = ((block_dst as i32) + block_delta) as usize;
+                }
+                row_dst += (row_pitch_in_bytes * 2) as usize;
+            }
+        }
+        return true;
+    }
+    pub fn unpack_dxt5a(&mut self, pDst: &mut [u8], row_pitch_in_bytes: u32, blocks_x: u32, blocks_y: u32, chunks_x: u32, chunks_y: u32) -> bool{
```
Comment (K0lb3) [inline src/crunch/crn_unpacker.rs:575]: bool -> Result

File: src/crunch/crn_unpacker.rs [hunk 13]
```diff
@@ -0,0 +1,742 @@
...
+                        pd_dst >>= 2;
+                    }
+                    block_dst = ((block_dst as i32) + block_delta) as usize;
+                }
+                row_dst += (row_pitch_in_bytes * 2) as usize;
+            }
+        }
+        return true;
+    }
+    pub fn unpack_dxn(&mut self, pDst: &mut [u8], row_pitch_in_bytes: u32, blocks_x: u32, blocks_y: u32, chunks_x: u32, chunks_y: u32) -> bool{
```
Comment (K0lb3) [inline src/crunch/crn_unpacker.rs:649]: bool -> Result

File: src/crunch/crn_utils.rs [hunk 14]
```diff
@@ -0,0 +1,57 @@
+#![allow(unused_parens)]
```
Comment (K0lb3) [inline src/crunch/crn_utils.rs:1]: let's just comment them out

File: src/crunch/crnlib.rs [hunk 15]
```diff
@@ -0,0 +1,422 @@
+#![allow(non_snake_case)]
+#![allow(non_camel_case_types)]
+#![allow(non_upper_case_globals)]
+#![allow(dead_code)]
```
Comment (K0lb3) [inline src/crunch/crnlib.rs:4]: as previously mentioned

File: src/crunch/crnlib.rs [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+    pub fn set_flag(&mut self, flag: u32, val: bool){
+        self.m_flags &= !flag;
+        if val {
+            self.m_flags |= flag
+        }
+    }
+
+   // Returns true if the input parameters are reasonable.
+   #[inline]
+    pub fn check(&mut self) -> bool {
```
Comment (K0lb3) [inline src/crunch/crnlib.rs:285]: bool -> Result

**Comment by K0lb3** [comment]
+1 all of that for unitycrunch.....

**Review by K0lb3 — CHANGES_REQUESTED** [review]
See comments

File: src/crunch.rs (hunk 2 above)
Comment (chayleaf) [reply src/crunch.rs]: or just `1.max(tex_info.m_width >> level_index)`

File: src/crunch/crn_decomp.rs (hunk 5 above)
Comment (chayleaf) [reply src/crunch/crn_decomp.rs]: and return keyword should be dropped

File: README.md
Comment (chayleaf) [inline README.md:1]: why is this changed so much? did you accidentally replace LF with CRLF or something?

File: resources/tests/README.md
Comment (chayleaf) [inline resources/tests/README.md:1]: same here, you seem to have broken the line endings, please try using git cli as it handles it correctly

File: src/crnlib.rs [hunk 17]
```diff
@@ -0,0 +1,144 @@
+use crate::crunch::crn_decomp::CrnHeader;
+
+// Supported compressed pixel formats.
+// Basically all the standard DX9 formats, with some swizzled DXT5 formats
+// (most of them supported by ATI's Compressonator), along with some ATI/X360 GPU specific formats.
+#[derive(PartialEq, PartialOrd)]
+#[repr(C)]
+pub enum CrnFormat {
+    CCrnfmtInvalid = -1,
```
Comment (chayleaf) [inline src/crnlib.rs:9]: drop `CCrnfmt` prefix from enum members. The reason it's like that in C is because enum members get added to global namespace. In Rust they are in the per-enum namespace, so adding prefixes like this is actually bad as it leads to duplication (`CrnFormat::CCrnfmtInvalid` vs `CrnFormat::Invalid`}

File: src/crnlib.rs [hunk 18]
```diff
@@ -0,0 +1,144 @@
+use crate::crunch::crn_decomp::CrnHeader;
+
+// Supported compressed pixel formats.
+// Basically all the standard DX9 formats, with some swizzled DXT5 formats
+// (most of them supported by ATI's Compressonator), along with some ATI/X360 GPU specific formats.
+#[derive(PartialEq, PartialOrd)]
+#[repr(C)]
```
Comment (chayleaf) [inline src/crnlib.rs:7]: `repr(C)` -> `repr(u32)`

File: src/crnlib.rs [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+
+    CCrnfmtEtc1,
+    CCrnfmtEtc2,
+    CCrnfmtEtc2a,
+    CCrnfmtEtc1s,
+    CCrnfmtEtc2as,
+
+    CCrnfmtTotal,
+
+    CCrnfmtForceDword = 0xFFFFFFFF
```
Comment (chayleaf) [inline src/crnlib.rs:41]: remove this in favor of `repr(u32)`

File: src/crnlib.rs [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+#[repr(C)]
+pub struct CrnTextureInfo {
+    pub m_struct_size: u32,
```
Comment (chayleaf) [inline src/crnlib.rs:46]: drop the `m_`. In C++ this is done to avoid confusion, in Rust it isn't needed as it's always clear what you refer to (e.g. no implicit `this->`)

File: src/crnlib.rs [hunk 21]
```diff
@@ -0,0 +1,144 @@
+use crate::crunch::crn_decomp::CrnHeader;
+
+// Supported compressed pixel formats.
+// Basically all the standard DX9 formats, with some swizzled DXT5 formats
+// (most of them supported by ATI's Compressonator), along with some ATI/X360 GPU specific formats.
+#[derive(PartialEq, PartialOrd)]
```
Comment (chayleaf) [inline src/crnlib.rs:6]: Add a `Default` derive, marking `Invalid` with `#[default]`

File: src/crnlib.rs [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+    pub m_height: u32,
+    pub m_levels: u32,
+    pub m_faces: u32,
+    pub m_bytes_per_block: u32,
+    pub m_userdata0: u32,
+    pub m_userdata1: u32,
+    pub m_format: CrnFormat
+}
+
+impl CrnTextureInfo{
```
Comment (chayleaf) [inline src/crnlib.rs:57]: after adding a `Default` derive for `CrnFormat`, replace this manual impl with `#[derive(Default)]`

File: src/crunch.rs [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+use core::cmp::max;
+use crate::bcn;
+extern crate alloc;
+
+struct CrunchDecodeHandler{
+    format: CrnFormat,
+    dxt_data: alloc::vec::Vec<u8>
+}
+
+fn crunch_unpack_level<'vec>(data: &[u8], data_size: u32, level_index: u32) -> Result<CrunchDecodeHandler, &'static str> {
```
Comment (chayleaf) [inline src/crunch.rs:17]: remove the unused lifetime

```suggestion
fn crunch_unpack_level(data: &[u8], data_size: u32, level_index: u32) -> Result<CrunchDecodeHandler, &'static str> {

```

File: src/crnlib.rs [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+    pub fn crnd_get_texture_info(&mut self, p_data: &[u8], data_size: u32) -> bool{
+        if data_size < core::mem::size_of::<CrnHeader>() as u32 {
+            return false;
+        }
+
+        if self.m_struct_size != core::mem::size_of::<CrnTextureInfo>() as u32{
+            return false;
+        }
+
+        let mut p_header: CrnHeader = CrnHeader::default();
```
Comment (chayleaf) [inline src/crnlib.rs:81]: `: CrnHeader` isn't needed here

```suggestion
        let mut p_header = CrnHeader::default();
```

File: src/crnlib.rs [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+            11 => CrnFormat::CCrnfmtEtc2,
+            12 => CrnFormat::CCrnfmtEtc2a,
+            13 => CrnFormat::CCrnfmtEtc1s,
+            14 => CrnFormat::CCrnfmtEtc2as,
+        
+            15 => CrnFormat::CCrnfmtTotal,
+        
+            0xFFFFFFFF => CrnFormat::CCrnfmtForceDword,
+
+            _ => CrnFormat::CCrnfmtInvalid
```
Comment (chayleaf) [inline src/crnlib.rs:126]: replace with `_ => return false,` and remove the if after this match

```suggestion
            _ => return false,
```

File: src/crunch.rs [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+    let mut tex_info: CrnTextureInfo = CrnTextureInfo::default();
+    if tex_info.crnd_get_texture_info(data, data_size) == false {
```
Comment (chayleaf) [inline src/crunch.rs:19]:
```suggestion
    if !tex_info.crnd_get_texture_info(data, data_size) {

```

File: src/crunch.rs [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+        return Err("Invalid crunch texture encoding.");
```
Comment (chayleaf) [inline src/crunch.rs:20]: In Rust, the convention is to have error messages in lowercase and with no trailing punctuation
```suggestion
        return Err("invalid crunch texture encoding");

```

Comment (chayleaf) [reply src/crunch.rs:20]: same goes for all other messages

File: src/crunch.rs [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+use crate::bcn;
+extern crate alloc;
+
+struct CrunchDecodeHandler{
+    format: CrnFormat,
+    dxt_data: alloc::vec::Vec<u8>
+}
+
+fn crunch_unpack_level<'vec>(data: &[u8], data_size: u32, level_index: u32) -> Result<CrunchDecodeHandler, &'static str> {
+    let mut tex_info: CrnTextureInfo = CrnTextureInfo::default();
```
Comment (chayleaf) [inline src/crunch.rs:18]:
```suggestion
    let mut tex_info = CrnTextureInfo::default();

```

File: src/crunch.rs [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+    if tex_info.crnd_get_texture_info(data, data_size) == false {
+        return Err("Invalid crunch texture encoding.");
+    }
+    if tex_info.m_faces != 1 {
+        // I think cubemaps have 6, but they are not in the same ballpark as Texture2D?
+        return Err("Texture2D must only have 1 number of faces.");
+    }
+    let mut p_context: crn_unpacker::CrnUnpacker<'_> = match crn_decomp::crnd_unpack_begin(&data, data_size){
```
Comment (chayleaf) [inline src/crunch.rs:26]:
```suggestion
    let mut p_context = match crn_decomp::crnd_unpack_begin(&data, data_size){

```

File: src/crunch.rs [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+        Ok(p_context) => p_context,
+        Err(res) => return Err(res)
+    };
+    let width = max(1, tex_info.m_width >> level_index);
```
Comment (chayleaf) [inline src/crunch.rs:30]: note that you fan also do `1.max(tex_info.m_width >> level_index)`

File: src/crunch.rs [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+    let height = max(1, tex_info.m_height >> level_index);
+    let blocks_x: u32 = max(1, ((width + 3) >> 2) as u32);
+    let blocks_y: u32 = max(1, ((height + 3) >> 2) as u32);
+    let row_pitch: u32 = blocks_x * match crn_decomp::crnd_get_bytes_per_dxt_block(&mut tex_info.m_format){
+        Ok(s) => s,
+        Err(e) => return Err(e)
+    };
+    let total_face_size: u32 = row_pitch * blocks_y;
```
Comment (chayleaf) [inline src/crunch.rs:38]: u32 type declarations here and above aren't needed

File: src/crunch/crn_consts.rs [hunk 32]
```diff
@@ -0,0 +1,107 @@
+// Moved Consts from structs.
```
Comment (chayleaf) [inline src/crunch/crn_consts.rs:1]: maybe drop `C_` prefix from const names?

File: src/crunch/crn_decomp.rs [hunk 33]
```diff
@@ -0,0 +1,224 @@
...
+            2 => u16::from_be_bytes([self.m_buf[0], self.m_buf[1]]) as u32,
+            3 => (self.m_buf[0] as u32) << 16 | u16::from_be_bytes([self.m_buf[1], self.m_buf[2]]) as u32,
+            4 => u32::from_be_bytes([self.m_buf[0], self.m_buf[1], self.m_buf[2], self.m_buf[3]]),
+            _ => panic!("Packed integer can hold a 4 byte buffer at max!")
+        }
+    }
+
+}
+
+impl<const N: usize> Default for CrnPackedUint<N>{
```
Comment (chayleaf) [inline src/crunch/crn_decomp.rs:62]: replace with `#[derive(Default)]`

File: src/crunch/crn_decomp.rs [hunk 34]
```diff
@@ -0,0 +1,224 @@
...
+//             m_color_selector_palette_entries: 0,
+//             m_alpha_endpoint_palette_entries: 0,
+//             m_alpha_selector_palette_entries: 0
+//         }
+//     }
+// }
+
+#[repr(C)]
+pub struct CrnPackedUint<const N: usize>{
+    pub m_buf: [u8; N]
```
Comment (chayleaf) [inline src/crunch/crn_decomp.rs:41]:
```suggestion
    pub buf: [u8; N]

```

**Review by chayleaf — CHANGES_REQUESTED** [review]
I only reviewed the first few files, but I have general recommendations for other files:

1. Use `rustfmt` for consistent formatting with other Rust code
2. Use `clippy`, it will point out many of the issues I told you about and extend it to the files I haven't checked.
3. Take care to fix all warnings (including ones printed by clippy), don't silence them

File: src/crnlib.rs (hunk 20 above)
Comment (chayleaf) [reply src/crnlib.rs:46]: same goes for all other structs

File: src/crnlib.rs (hunk 17 above)
Comment (chayleaf) [reply src/crnlib.rs:9]: In fact, many codebases would have this called `Format` and refer to it as `crn::Format::Invalid`

File: src/crunch/crn_decomp.rs (hunk 33 above)
Comment (K0lb3) [reply src/crunch/crn_decomp.rs:62]: Rust's Default trait does not support generic constants, so it can't be replaced with a derive.

File: src/crnlib.rs [hunk 35]
```diff
@@ -0,0 +1,140 @@
...
+
+            10 => CrnFormat::Etc1,
+            11 => CrnFormat::Etc2,
+            12 => CrnFormat::Etc2a,
+            13 => CrnFormat::Etc1s,
+            14 => CrnFormat::Etc2as,
+
+            15 => CrnFormat::Total,
+
+            _ => CrnFormat::Invalid,
```
Comment (chayleaf) [inline src/crnlib.rs:121]:
```suggestion
```suggestion
            _ => CrnFormat::Invalid,
```

**Review by K0lb3 — APPROVED** [review]
